File: macos/Shared/S3Store.swift
import Foundation
import OSLog

private let log = os.Logger(subsystem: "com.toots.tsync", category: "S3Store")

/// Bundles S3Client + Config + domainName with domain-aware key construction and S3 operations.
public struct S3Store: Sendable {
    public let client: S3Client
    public let config: Config
    public let domainName: String

    public init(client: S3Client, config: Config, domainName: String) {
        self.client = client
        self.config = config
        self.domainName = domainName
    }

    // MARK: Key helpers

    public var domainPrefix: String { config.domainPrefix(domainName) }
    public var chunkPrefix: String { config.chunkPrefix() }
    public var journalPrefix: String { config.journalPrefix(domainName) }
    public var versionKey: String { config.versionKey(domainName) }

    public func key(for relativePath: String) -> String { domainPrefix + relativePath }

    public func relativePath(of key: String) -> String {
        key.hasPrefix(domainPrefix) ? String(key.dropFirst(domainPrefix.count)) : key
    }

    public func isDirectoryKey(_ key: String) -> Bool { key.hasSuffix("/") }
    public func directoryKey(_ key: String) -> String { key.hasSuffix("/") ? key : key + "/" }

    public func name(of key: String) -> String {
        key.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
            .components(separatedBy: "/").last ?? key
    }

    // MARK: Domain-aware S3 operations

    public func upload(key: String, from url: URL) async throws {
        log.info("upload: \(key, privacy: .public)")
        try await client.putChunked(key: key, fileURL: url, chunkPrefix: chunkPrefix)
    }

    public func download(key: String, to url: URL) async throws {
        log.info("download: \(key, privacy: .public)")
        try await client.getChunked(key: key, to: url, chunkPrefix: chunkPrefix)
    }

    public func head(key: String) async throws -> (size: Int64, modified: Date?, etag: String?) {
        let (size, modified, etag, contentType) = try await client.head(key: key)
        guard contentType == ChunkManifest.contentType,
              let data = try? await client.getData(key: key),
              let manifest = try? JSONDecoder().decode(ChunkManifest.self, from: data)
        else { return (size, modified, etag) }
        return (manifest.size, modified, etag)
    }

    public func createDirectory(key: String) async throws {
        let dirKey = directoryKey(key)
        log.info("createDirectory: \(dirKey, privacy: .public)")
        try await client.putEmpty(key: dirKey)
    }

    public func listDirectory(prefix: String) async throws -> (dirs: [String], files: [String]) {
        let (subPrefixes, keys) = try await client.list(prefix: prefix, delimiter: "/")
        var dirs = subPrefixes
        for key in keys where key.hasSuffix("/") && !dirs.contains(key) {
            dirs.append(key)
        }
        return (dirs, keys.filter { !$0.hasSuffix("/") })
    }

    public func delete(key: String) async throws {
        log.info("delete: \(key, privacy: .public)")
        if isDirectoryKey(key) {
            let objects = try await client.listWithMetadata(prefix: key)
            guard !objects.isEmpty else { return }
            if config.versioning {
                let files = objects.filter { !$0.key.hasSuffix("/") }
                try await withBoundedConcurrency(files, maxConcurrent: 8) { obj in
                    let trash = Versioning.trashKey(for: obj.key, store: self)
                    try await self.client.copy(from: obj.key, to: trash)
                }
            }
            try await client.deleteObjects(keys: objects.map(\.key))
        } else {
            if config.versioning {
                let trash = Versioning.trashKey(for: key, store: self)
                try await client.copy(from: key, to: trash)
            }
            try await client.delete(key: key)
        }
    }

    // MARK: Journal

    /// Writes a journal entry and bumps the version file. Fire-and-forget — logs errors, never throws.
    /// Pass a pre-generated entryKey when the key was recorded locally before the S3 op.
    public func writeJournal(ops: [JournalOp], entryKey: String? = nil) async {
        let filename = entryKey ?? Journal.entryKey()
        let key = journalPrefix + filename
        let data = Journal.encode(ops)
        do {
            try await client.putData(key: versionKey, data: Data(filename.utf8), contentType: "text/plain")
            try await client.putData(key: key, data: data, contentType: "application/x-ndjson")
        } catch {
            log.error("writeJournal: \(error, privacy: .public)")
        }
    }

    /// Returns the current version string (latest journal entry filename), or nil if none yet.
    public func fetchVersion() async throws -> String? {
        guard let data = try? await client.getData(key: versionKey) else { return nil }
        return String(data: data, encoding: .utf8)
    }

    /// Replays any journal entries that were written locally but not yet committed to S3
    /// (i.e. the extension crashed between writing the local pending file and completing
    /// the S3 operation). Skips ops where a foreign client made a newer change to the
    /// same key. Call at the start of `tsync sync`.
    public func recoverPendingOps(mountURL: URL) async throws {
        let myUUID = Journal.clientUUID()
        let pending = Journal.localPendingEntries(forUUID: myUUID)
        guard !pending.isEmpty else { return }

        let remoteKeyFilenames = Set(
            try await client.listKeys(prefix: journalPrefix)
                .map { $0.components(separatedBy: "/").last ?? $0 }
        )

        for (entryKey, ops) in pending {
            // Crashed after S3 write, before local delete — just clean up.
            if remoteKeyFilenames.contains(entryKey) {
                Journal.deleteLocalPending(entryKey: entryKey)
                continue
            }

            // Collect keys touched by foreign entries newer than this pending entry.
            let entryMs = Journal.timestampMs(fromFilename: entryKey)
            let hasNewerForeign = remoteKeyFilenames.contains {
                Journal.timestampMs(fromFilename: $0) > entryMs &&
                Journal.clientUUID(fromFilename: $0) != myUUID
            }
            var remotelyModifiedKeys = Set<String>()
            if hasNewerForeign {
                let foreignEntries = try await listJournal(startAfter: journalPrefix + entryKey)
                    .filter { $0.clientUUID != myUUID }
                for entry in foreignEntries {
                    for op in entry.ops {
                        remotelyModifiedKeys.insert(op.key)
                        if let src = op.src { remotelyModifiedKeys.insert(src) }
                    }
                }
            }

            var replayed: [JournalOp] = []
            for op in ops where !remotelyModifiedKeys.contains(op.key) {
                do {
                    switch op.op {
                    case "put":
                        let localFile = mountURL.appendingPathComponent(op.key)
                        guard FileManager.default.fileExists(atPath: localFile.path) else { continue }
                        try await upload(key: key(for: op.key), from: localFile)
                        replayed.append(op)
                    case "delete":
                        try await delete(key: key(for: op.key))
                        replayed.append(op)
                    case "mkdir":
                        try await createDirectory(key: key(for: op.key))
                        replayed.append(op)
                    case "rmdir":
                        try await delete(key: directoryKey(key(for: op.key)))
                        replayed.append(op)
                    case "rename":
                        if let src = op.src {
                            let srcKey = key(for: src)
                            let dstKey = key(for: op.key)
                            if isDirectoryKey(src) {
                                try await renameDirectory(from: directoryKey(srcKey), to: directoryKey(dstKey))
                            } else {
                                try await copy(from: srcKey, to: dstKey)
                                try await delete(key: srcKey)
                            }
                            replayed.append(op)
                        }
                    default: break
                    }
                } catch {
                    log.error("recovery: failed to replay \(op.op, privacy: .public) \(op.key, privacy: .public): \(error, privacy: .public)")
                }
            }

            if !replayed.isEmpty {
                await writeJournal(ops: replayed, entryKey: entryKey)
                log.info("recovery: replayed \(replayed.count, privacy: .public) op(s) from \(entryKey, privacy: .public)")
            }
            Journal.deleteLocalPending(entryKey: entryKey)
        }
    }

    /// Lists journal entries after startAfter (nil = from the beginning).
    public func listJournal(startAfter: String?) async throws -> [JournalEntry] {
        let keys = try await client.listKeys(prefix: journalPrefix, startAfter: startAfter)
        var entries: [JournalEntry] = []
        for key in keys {
            guard let data = try? await client.getData(key: key) else { continue }
            let filename = key.components(separatedBy: "/").last ?? key
            entries.append(JournalEntry(
                s3Key: key,
                clientUUID: Journal.clientUUID(fromFilename: filename),
                timestampMs: Journal.timestampMs(fromFilename: filename),
                ops: Journal.decode(data)
            ))
        }
        return entries
    }

    public func copy(from fromKey: String, to toKey: String) async throws {
        try await client.copy(from: fromKey, to: toKey)
    }

    /// Recursively moves all S3 objects under oldPrefix to newPrefix.
    /// Used for directory rename — does not create trash copies (it's a move, not a delete).
    public func renameDirectory(from oldPrefix: String, to newPrefix: String) async throws {
        let objects = try await client.listWithMetadata(prefix: oldPrefix)
        log.info("renameDirectory: \(oldPrefix, privacy: .public) → \(newPrefix, privacy: .public) (\(objects.count, privacy: .public) objects)")
        try await withBoundedConcurrency(objects, maxConcurrent: 8) { obj in
            let newKey = newPrefix + String(obj.key.dropFirst(oldPrefix.count))
            try await self.client.copy(from: obj.key, to: newKey)
        }
        try await client.deleteObjects(keys: objects.map(\.key))
    }
}

File: linux/lib/s3_store.ml
type t = {
  client : S3_client.t;
  domain_name : string;
  domain_prefix : string;
  chunk_prefix : string;
  trash_prefix : string;
  versioning : bool;
  journal_prefix : string;
  version_key : string;
}

let make ~client ~domain_name ~domain_prefix ~chunk_prefix ~trash_prefix
    ~versioning ~journal_prefix ~version_key =
  {
    client;
    domain_name;
    domain_prefix;
    chunk_prefix;
    trash_prefix;
    versioning;
    journal_prefix;
    version_key;
  }

let upload t ~key ~src_path =
  S3_client.put_chunked t.client ~key ~src_path ~chunk_prefix:t.chunk_prefix

let download t ~key ~dst_path =
  Cache.ensure_parent_dir dst_path;
  S3_client.get_chunked t.client ~key ~dst_path ~chunk_prefix:t.chunk_prefix

let delete_file t ~key =
  if t.versioning then begin
    let trash_key =
      Versioning.trash_key ~s3_key:key ~domain_prefix:t.domain_prefix
        ~trash_prefix:t.trash_prefix
    in
    S3_client.copy t.client ~src_key:key ~dst_key:trash_key ()
  end;
  S3_client.delete t.client ~key ()

(* Delete a directory recursively without versioning (directory markers are not trashed) *)
let delete_dir t ~prefix =
  let all = S3_client.list_all t.client ~prefix () in
  let keys = List.map (fun e -> e.S3_client.key) all in
  S3_client.delete_multi t.client keys

let create_directory t ~key =
  S3_client.put t.client ~content_type:"application/x-directory" ~key ~data:""
    ()

let rename_file t ~src_key ~dst_key =
  S3_client.copy t.client ~src_key ~dst_key ();
  S3_client.delete t.client ~key:src_key ()

let rename_directory t ~src_prefix ~dst_prefix =
  let all = S3_client.list_all t.client ~prefix:src_prefix () in
  let src_len = String.length src_prefix in
  List.iter
    (fun e ->
      let suffix =
        String.sub e.S3_client.key src_len
          (String.length e.S3_client.key - src_len)
      in
      S3_client.copy t.client ~src_key:e.S3_client.key
        ~dst_key:(dst_prefix ^ suffix) ())
    all;
  S3_client.delete_multi t.client (List.map (fun e -> e.S3_client.key) all)

let list_directory t ~prefix = S3_client.list_directory t.client ~prefix ()
let head_opt t ~key = S3_client.head_opt t.client ~key ()

(* Like head_opt but returns the real file size for chunked manifests *)
let stat_file t ~key =
  match S3_client.head_opt t.client ~key () with
    | None -> None
    | Some c when c.S3_client.content_type = Some Chunk_manifest.content_type ->
        (try
           let m = Chunk_manifest.of_string (S3_client.get t.client ~key ()) in
           Some { c with S3_client.size = Int64.to_int m.size }
         with _ -> Some c)
    | Some c -> Some c
let domain_name t = t.domain_name
let domain_prefix t = t.domain_prefix
let journal_prefix t = t.journal_prefix

(* ── Journal ─────────────────────────────────────────────────────────────── *)

(* Write journal entry to S3 only; returns the entry key used.
   The version key is NOT updated — call bump_version separately. *)
let write_journal_entry ?entry_key ops t =
  let ek = match entry_key with Some k -> k | None -> Journal.entry_key () in
  let key = t.journal_prefix ^ ek in
  S3_client.put t.client ~content_type:"application/x-ndjson" ~key
    ~data:(Journal.encode ops) ();
  ek

(* Update the version key to point to a given entry key. *)
let bump_version t entry_key =
  S3_client.put t.client ~key:t.version_key ~data:entry_key ()

(* Write journal entry then bump version. Used for crash recovery only, not the hot path. *)
let write_journal ?entry_key ops t =
  let ek = match entry_key with Some k -> k | None -> Journal.entry_key () in
  ignore (write_journal_entry ~entry_key:ek ops t);
  bump_version t ek

let fetch_version t =
  match S3_client.head_opt t.client ~key:t.version_key () with
    | None -> None
    | Some _ ->
        Some (String.trim (S3_client.get t.client ~key:t.version_key ()))

(* Returns (entry_key_basename, client_uuid) list, optionally filtered by start_after.
   start_after may be a bare basename or a full S3 key (Filename.basename is applied). *)
let list_journal_keys ?start_after t () =
  let all = S3_client.list_all t.client ~prefix:t.journal_prefix () in
  let prefix_len = String.length t.journal_prefix in
  let sa_base = Option.map Filename.basename start_after in
  List.filter_map
    (fun (e : S3_client.file_entry) ->
      let basename =
        if String.length e.key > prefix_len then
          String.sub e.key prefix_len (String.length e.key - prefix_len)
        else e.key
      in
      match sa_base with
        | Some sa when basename <= sa -> None
        | _ -> (
            try
              ignore (Journal.timestamp_ms_of_filename basename);
              Some (basename, Journal.client_uuid_of_filename basename)
            with _ -> None))
    all

let get_journal_entry t entry_key =
  let key = t.journal_prefix ^ entry_key in
  try Some (Journal.decode (S3_client.get t.client ~key ()))
  with _ -> None

(* Replay locally-pending WAL entries that didn't make it to S3 before crash *)
let recover_pending_ops t =
  let my_uuid = Journal.client_uuid () in
  List.iter
    (fun (entry_key, ops) ->
      let remote_key = t.journal_prefix ^ entry_key in
      if S3_client.head_opt t.client ~key:remote_key () <> None then
        (* already made it to S3 before crash *)
        Journal.delete_local_pending ~entry_key
      else begin
        let newer_keys = list_journal_keys ~start_after:entry_key t () in
        let remotely_modified = Hashtbl.create 16 in
        List.iter
          (fun (ek, uuid) ->
            if uuid <> my_uuid then
              match get_journal_entry t ek with
                | None -> ()
                | Some remote_ops ->
                    List.iter
                      (fun op ->
                        match op with
                          | `Put (k, _) | `Delete k | `Mkdir k | `Rmdir k ->
                              Hashtbl.replace remotely_modified k ()
                          | `Rename (k, src, _) ->
                              Hashtbl.replace remotely_modified k ();
                              Hashtbl.replace remotely_modified src ())
                      remote_ops)
          newer_keys;
        let replayed =
          List.filter
            (fun op ->
              let k =
                match op with
                  | `Put (k, _) | `Delete k | `Mkdir k | `Rmdir k -> k
                  | `Rename (k, _, _) -> k
              in
              not (Hashtbl.mem remotely_modified k))
            ops
        in
        List.iter
          (fun op ->
            try
              match op with
                | `Put (rel_key, _) ->
                    let full_key = t.domain_prefix ^ rel_key in
                    let cache_path =
                      Cache.cache_path ~domain_name:t.domain_name
                        ~domain_prefix:t.domain_prefix full_key
                    in
                    if Sys.file_exists cache_path then
                      ignore (upload t ~key:full_key ~src_path:cache_path)
                | `Delete rel_key ->
                    delete_file t ~key:(t.domain_prefix ^ rel_key)
                | `Mkdir rel_key ->
                    create_directory t ~key:(t.domain_prefix ^ rel_key)
                | `Rmdir rel_key ->
                    delete_dir t ~prefix:(t.domain_prefix ^ rel_key)
                | `Rename (dst_rel, src_rel, _) ->
                    let src_key = t.domain_prefix ^ src_rel in
                    let dst_key = t.domain_prefix ^ dst_rel in
                    (match
                       S3_client.head_opt t.client ~key:(src_key ^ "/") ()
                     with
                      | Some _ ->
                          rename_directory t ~src_prefix:(src_key ^ "/")
                            ~dst_prefix:(dst_key ^ "/")
                      | None -> rename_file t ~src_key ~dst_key)
            with exn ->
              Log.err "recover_pending_ops: %s" (Printexc.to_string exn))
          replayed;
        if replayed <> [] then write_journal ~entry_key replayed t;
        Journal.delete_local_pending ~entry_key
      end)
    (Journal.local_pending_entries ~uuid:my_uuid)
